File: daysPRO/EditViewController.m
//
//  EditViewController.m
//  Days Pro
//
//  Created by Oliver Kulpakko on 2017-01-15.
//  Copyright © 2017 East Studios. All rights reserved.
//

#import "EditViewController.h"

@interface EditViewController ()

@end

@implementation EditViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    if (self.isEditing) {
        self.title = self.event.name;
        self.datePicker.date = self.event.startDate;
        self.nameTextField.text = self.event.name;
    } else {
        self.datePicker.minimumDate = [NSDate date];
    }
    
    UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(dismissKeyboards)];
    [self.view addGestureRecognizer:tap];
    
    self.eventImageView.image = self.event.image;
    self.eventImageView.contentMode = UIViewContentModeScaleAspectFill;
    self.eventImageView.clipsToBounds = true;

    [self.navigationController.navigationBar setTitleTextAttributes:
    @{NSForegroundColorAttributeName:[ThemeManager getThemeColor]}];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)dismissKeyboards {
    [self.nameTextField resignFirstResponder];
}

- (IBAction)saveEvent:(id)sender {
    if (self.nameTextField.text.length == 0) {
        [SVProgressHUD showErrorWithStatus:NSLocalizedString(@"No Name", nil)];
    } else {
        NSDate *date = [self dateWithHour:0 minute:0 second:0 date:self.datePicker.date];
        if (self.isEditing) {
            [[DataManager sharedManager] updateEvent:self.event withName:self.nameTextField.text
                                           startDate:date endDate:date details:nil image:self.eventImageView.image];
            [[DataManager sharedManager] saveContext];
            
            [Answers logCustomEventWithName:@"Edit event" customAttributes:@{@"Name":_nameTextField.text}];
        } else {
            [[DataManager sharedManager] createEventWithName:self.nameTextField.text
                                                   startDate:date endDate:date details:nil image:self.eventImageView.image];
            
            [[DataManager sharedManager] saveContext];
            [Answers logCustomEventWithName:@"Add event" customAttributes:@{@"Name":_nameTextField.text}];
        }
        
        [self dismissViewControllerAnimated:YES completion:nil];
        [SVProgressHUD showSuccessWithStatus:nil];
    }
}

- (IBAction)cancel:(id)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (IBAction)changeImage:(id)sender {
    UIImagePickerController *picker;
    
    picker = [[UIImagePickerController alloc] init];
    picker.delegate = self;
    picker.allowsEditing = true;
    UIAlertController *alertController = [UIAlertController
                                          alertControllerWithTitle:NSLocalizedString(@"Add Image", nil)
                                          message:nil
                                          preferredStyle:UIAlertControllerStyleAlert];
    
    UIAlertAction *takePicture = [UIAlertAction
                                  actionWithTitle:NSLocalizedString(@"Take a Picture", nil)
                                  style:UIAlertActionStyleDefault
                                  handler:^(UIAlertAction *action)
                                  {
                                      picker.sourceType = UIImagePickerControllerSourceTypeCamera;
                                      picker.allowsEditing = false;
                                      [self presentViewController:picker animated:YES completion:nil];
                                  }];
    
    UIAlertAction *cameraRoll = [UIAlertAction
                                 actionWithTitle:NSLocalizedString(@"Select a Picture", nil)
                                 style:UIAlertActionStyleDefault
                                 handler:^(UIAlertAction *action)
                                 {
                                     picker.sourceType = UIImagePickerControllerSourceTypePhotoLibrary;
                                     [self presentViewController:picker animated:YES completion:^{
                                         picker.topViewController.title = NSLocalizedString(@"Select a Picture", nil);
                                         picker.navigationBar.translucent = NO;
                                         picker.allowsEditing = false;
                                         picker.navigationBar.barStyle = UIBarStyleDefault;
                                         [picker setNavigationBarHidden:NO animated:NO];
                                     }];
                                 }];
    
    UIAlertAction *removeImage = [UIAlertAction
                                  actionWithTitle:NSLocalizedString(@"Remove Image", nil)
                                  style:UIAlertActionStyleDefault
                                  handler:^(UIAlertAction *action)
                                  {
                                      //remove the image
                                      NSFileManager *fileManager = [NSFileManager defaultManager];
                                      NSString *documentsPath = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) objectAtIndex:0];
                                      
                                      NSString *filePath = [documentsPath stringByAppendingPathComponent:self.event.uuid];
                                      NSError *error;
                                      [fileManager removeItemAtPath:filePath error:&error];
                                      self.eventImageView.image = nil;
                                  }];
    
    UIAlertAction *cancel = [UIAlertAction
                             actionWithTitle:NSLocalizedString(@"Cancel", nil)
                             style:UIAlertActionStyleCancel
                             handler:nil];
    [alertController addAction:takePicture];
    [alertController addAction:cameraRoll];
    
    //Only show the remove button if there's an image
    if (self.eventImageView.image) {
        [alertController addAction:removeImage];
    }
    
    [alertController addAction:cancel];
    [self presentViewController:alertController animated:YES completion:nil];
}

-(NSDate *)dateWithHour:(NSInteger)hour minute:(NSInteger)minute second:(NSInteger)second date:(NSDate *)date {
    NSCalendar *calendar = [NSCalendar currentCalendar];
    NSDateComponents *components = [calendar components:
                                    NSCalendarUnitYear| NSCalendarUnitMonth| NSCalendarUnitDay fromDate:date];
    [components setHour:hour];
    [components setMinute:minute];
    [components setSecond:second];
    NSDate *newDate = [calendar dateFromComponents:components];
    return newDate;
}

#pragma mark - ImagePickerController Delegate
- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingImage:(UIImage *)image editingInfo:(NSDictionary *)editingInfo {
    [picker dismissViewControllerAnimated:YES completion:nil];
    self.eventImageView.image = image;
}

- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker {
    [picker dismissViewControllerAnimated:YES completion:nil];
}

@end

File: daysPRO/EditViewController.swift
//
//  EditViewController.swift
//  Days Pro
//
//  Created by Oliver Kulpakko on 2017-05-24.
//  Copyright © 2017 Oliver Kulpakko. All rights reserved.
//

import UIKit
import SVProgressHUD
import ImagePicker

class EditViewController: UIViewController, ImagePickerDelegate {
    var event: Event?
    
    @IBOutlet weak var timePicker: UIDatePicker!
    @IBOutlet weak var datePicker: UIDatePicker!
    @IBOutlet weak var dateTimeSwitch: UISegmentedControl!
    @IBOutlet weak var descriptionTextField: UITextField!
    @IBOutlet weak var nameTextField: UITextField!
    @IBOutlet weak var imageViewButton: UIButton!
    @IBOutlet weak var imageView: UIImageView!
    @IBOutlet weak var cancelButton: UIBarButtonItem!
    @IBOutlet weak var saveButton: UIBarButtonItem!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.hideKeyboardWhenTappedAround()
        if let currentEvent = self.event {
            self.datePicker.date = currentEvent.date()
            self.timePicker.date = self.time(date: currentEvent.date())
            
            self.nameTextField.text = currentEvent.name
            self.descriptionTextField.text = currentEvent.details
            self.title = currentEvent.name
            
            self.imageView.image = currentEvent.image()
        }
    }
    
    @IBAction func changeBetweenDateAndTime(_ sender: Any) {
        self.datePicker.isHidden = !self.datePicker.isHidden
        self.timePicker.isHidden = !self.timePicker.isHidden
    }

    @IBAction func cancel(_ sender: Any) {
        self.dismiss(animated: true, completion: nil)
    }
    
    @IBAction func save(_ sender: Any) {
        if self.nameTextField.text == nil {
            SVProgressHUD.showError(withStatus: NSLocalizedString("No Name", comment: ""))
            return
        }
        if (self.event != nil) {
            DataManager.shared().update(event, withName: nameTextField.text, date: self.combinedDate(), details: self.descriptionTextField.text, image: self.imageView.image)
            DataManager.shared().saveContext()
        } else {
            DataManager.shared().createEvent(withName: nameTextField.text, date: self.combinedDate(), details: self.descriptionTextField.text, image: self.imageView.image)
            DataManager.shared().saveContext()
        }
        self.cancel(sender)
    }
    
    @IBAction func selectImage(_ sender: Any) {
        let imagePickerController = ImagePickerController()
        imagePickerController.delegate = self
        imagePickerController.imageLimit = 1
        present(imagePickerController, animated: true, completion: nil)
    }
    
    func combinedDate() -> Date? {
        let calendar = NSCalendar.current
        
        let dateComponents = calendar.dateComponents([.year, .month, .day], from: self.datePicker.date)
        let timeComponents = calendar.dateComponents([.hour, .minute, .second], from: self.timePicker.date)
        
        var mergedComponments = DateComponents()
        mergedComponments.year = dateComponents.year!
        mergedComponments.month = dateComponents.month!
        mergedComponments.day = dateComponents.day!
        mergedComponments.hour = timeComponents.hour!
        mergedComponments.minute = timeComponents.minute!
        mergedComponments.second = timeComponents.second!
        
        return calendar.date(from: mergedComponments)
    }
    
    func time(date: Date) -> Date {
        let calendar = Calendar.current

        var dateComponents = DateComponents()
        dateComponents.hour = calendar.component(.hour, from: date)
        dateComponents.minute = calendar.component(.minute, from: date)
        dateComponents.second = calendar.component(.second, from: date)

        return calendar.date(from: dateComponents)!
    }
    
    func cancelButtonDidPress(_ imagePicker: ImagePickerController) {
        imagePicker.dismiss(animated: true, completion: nil)
    }
    
    func doneButtonDidPress(_ imagePicker: ImagePickerController, images: [UIImage]) {
        self.imageView.image = images.first
        imagePicker.dismiss(animated: true, completion: nil)
    }
    
    func wrapperDidPress(_ imagePicker: ImagePickerController, images: [UIImage]) {
        self.imageView.image = images.first
        imagePicker.dismiss(animated: true, completion: nil)
    }
}

extension UIViewController {
    func hideKeyboardWhenTappedAround() {
        let tap: UITapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(UIViewController.dismissKeyboard))
        tap.cancelsTouchesInView = false
        view.addGestureRecognizer(tap)
    }
    
    func dismissKeyboard() {
        view.endEditing(true)
    }
}
